### PR TITLE
Add Lot No to Select Spool Modal

### LINKED ISSUE #52: [Request] Lot no or some other identifier in UI
### Is your feature request related to a problem? Please describe.

Have multiple spools of same color from same manufacturer but no idenfier availiable for spool on octoprint

### Describe the solution you'd like

SOme identifer like Lot No visible on the UI to better identify spool being used

### Describe alternatives you've considered

I check remaining weight on the spools and match using spoolman UI to see which lot no has now much remaininig

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jameswolff96**
## Description
This PR adds ~the Spool ID and the Lot Number to the Sidebar and~ adds the Lot Number to the Select Spool Modal

Edit (by @mdziekon): Sidebar modifications moved out of this PR.

## Related Issue / Discussion
Fixes #52 

## How has this been tested?
I tested with lot numbers between 10 and 64 characters (the maximum allowed by Spoolman)

## Screenshots (if appropriate):
image
image
image

## Types of changes
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have verified that my changes do not break the overall functionality of the plugin.

### CODE REVIEW

**Comment by jameswolff96** [comment]
This was the cleanest I was able to come up with, I am definitely open to suggestions, though!

File: octoprint_Spoolman/static/js/Spoolman_modal_selectSpool.js
Comment (mdziekon) [inline octoprint_Spoolman/static/js/Spoolman_modal_selectSpool.js:153]: To move forward with this functionality, I think it has to become configurable via the plugin's settings panel. For people like myself, who don't print that much, using lot number is completely unnecessary, therefore this column will just waste screen space to display no value.

Some of the infrastructure to show / hide columns is already there, so hopefully this will boil down to just adding a checkbox in the settings and that should be all. Ideally, all of the currently present columns should be toggleable, but as that's not the main focus of this PR, I'm fine with introducing it for just this new column.

File: octoprint_Spoolman/templates/Spoolman_sidebar.jinja2
Comment (mdziekon) [inline octoprint_Spoolman/templates/Spoolman_sidebar.jinja2:117]: As spool id is dynamic in length, ids longer than just 1 character seem to shift the layout of spool row display. As much as I would love to prevent too much vertical space being wasted, I believe it will be better to move it to the "column" with the text, and display it eg. below all the current data, in a separate line.

File: octoprint_Spoolman/templates/Spoolman_sidebar.jinja2
Comment (mdziekon) [inline octoprint_Spoolman/templates/Spoolman_sidebar.jinja2:133]: Putting the lot number before the "title" (name & brand of the filament) seems odd to me. Similar to the id proposal from the previous comment, I think it would be better to put it below, either on the same line as id, or one line below.

Comment (mdziekon) [inline octoprint_Spoolman/templates/Spoolman_sidebar.jinja2:133]: I'm also wondering - how important it is to display "lot no" its entirety? Potential 64 characters seems really long, and the space seems quite crowded already, with just the filament type, name & brand.

I was thinking, that maybe the lot number could be treated a bit differently - make it truncated (eg. for lot numbers longer than 9 characters, show the first 3 and last 3 characters, with `...` in the middle), and retain the full value in a tooltip, so that it's still accessible when hovered over? It's not that important in the list picker, but here, in the sidebar, it might be worth the effort, as long as it still serves its purpose. What do you think about that?

**Review by mdziekon — CHANGES_REQUESTED** [review]
Quite tricky PR, in terms of visual design, so I've left some comments to improve some things and maybe discuss different approaches that might work better.

The sidebar enhancement feature might take some time to iron it out, so I strongly suggest to split that PR into two separate ones - one for the picker, and one for the sidebar. I think the picker should be fairly easy to improve to a "mergeable" state, so should get merged faster. Both need some shared code (the typings), but since you've already split that into nice commits history, that shouldn't be a problem to sync things up when necessary.

As always, thank you for your contribution, and if you have any questions or other remarks, feel free to respond to the in-code comments.

File: octoprint_Spoolman/static/js/Spoolman_modal_selectSpool.js
Comment (jameswolff96) [reply octoprint_Spoolman/static/js/Spoolman_modal_selectSpool.js:153]: I will take a look at how the settings integration works, truthfully this is the first Octoprint plugin I have worked on. I will take a look at Octoprint's documentation, but any pointers for your implementation would be appreciated!

I would obviously need to add a section to `Spoolman_settings.jinja2`, but does it need to be in a new tab, for example `Display Settings`?

**Comment by jameswolff96** [comment]
> The sidebar enhancement feature might take some time to iron it out, so I strongly suggest to split that PR into two separate ones - one for the picker, and one for the sidebar. I think the picker should be fairly easy to improve to a "mergeable" state, so should get merged faster. Both need some shared code (the typings), but since you've already split that into nice commits history, that shouldn't be a problem to sync things up when necessary.

I appreciate the comments! I will gladly split this into two PRS, am I ok to force push and delete the last commit, or would you like me to revert the changes locally and push a new commit to preserve the history?

File: octoprint_Spoolman/static/js/Spoolman_modal_selectSpool.js
Comment (jameswolff96) [reply octoprint_Spoolman/static/js/Spoolman_modal_selectSpool.js:153]: I was able to get the setting working with it defaulting to false, however changing the setting requires a page reload before it takes effect. I imagine there is a way to force that update, possibly in `Spoolman_settings.js`? Here is the updated code for the table visibility for reference.
image

Comment (mdziekon) [reply octoprint_Spoolman/static/js/Spoolman_modal_selectSpool.js:153]: > I would obviously need to add a section to `Spoolman_settings.jinja2`, but does it need to be in a new tab, for example `Display Settings`?

Yes, that's correct, that's where all settings reside. Creating a new tab for the display settings is a good idea, so feel free to explore that, but I won't block the PR if you won't be able to implement it that way.

Comment (mdziekon) [reply octoprint_Spoolman/static/js/Spoolman_modal_selectSpool.js:153]: > I was able to get the setting working with it defaulting to false, however changing the setting requires a page reload before it takes effect. I imagine there is a way to force that update, possibly in `Spoolman_settings.js`? Here is the updated code for the table visibility for reference.

Octoprint uses the Knockout framework which implements [observables](https://knockoutjs.com/documentation/observables.html) to handle automatic updates like the one you would like to see. I'm not super versed in that framework myself, but I think you would have to turn `tableAttributeVisibility` variable into an observable (maybe `ko.computed`) to auto update on settings change. To be fair, I don't know how Knockout "knows" that a dependency of such computed observable has changed, but you could give it a try, there should be at least one usage of `ko.computed` somewhere in the codebase.

**Comment by mdziekon** [comment]
> I appreciate the comments! I will gladly split this into two PRS, am I ok to force push and delete the last commit, or would you like me to revert the changes locally and push a new commit to preserve the history?

Whatever works best for you I think. Once ready for merge, I'll perform a squash merge anyway, so the individual commits will be lost anyway. My suggestion about reusing some commits was mainly for reuse across your two separate PRs, but ultimately one of the two (PRs) will have to be rebased anyway, but I think using shared commit will make rebasing potentially a bit easier for you.

File: octoprint_Spoolman/templates/Spoolman_sidebar.jinja2
Comment (jameswolff96) [reply octoprint_Spoolman/templates/Spoolman_sidebar.jinja2]: I had not noticed that, but after adding a couple hundred spools to my test instance I see what you mean, I will explore this a bit more in the new PR.

File: octoprint_Spoolman/templates/Spoolman_sidebar.jinja2
Comment (jameswolff96) [reply octoprint_Spoolman/templates/Spoolman_sidebar.jinja2]: The decision here was based on the particular filament I was testing with preforming a line break after the filament type, I thought there may have been a purpose for this, and so thought there would be a good place to fill some unused space, however I do agree that below all of this information is a much better suited place for this. I will explore this in the other PR.

Comment (jameswolff96) [reply octoprint_Spoolman/templates/Spoolman_sidebar.jinja2]: I like this idea, I will explore it some more in a separate PR!

**Review by mdziekon — APPROVED** [review]

**Comment by mdziekon** [comment]
@jameswolff96 thank you for your contribution, and looking forward to the follow-up PR with the Sidebar changes! :tada: 

BTW, I'll wait for a bit before preparing the next release to give you some time to work on the said PR.

**Pull request merged**
